zookeeper/src/main/java/com/linecorp/armeria/client/zookeeper/NodeChildEndpointGroup.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.zookeeper;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.apache.zookeeper.ZooKeeper;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.endpoint.EndpointGroup;
+import com.linecorp.armeria.common.zookeeper.DefaultNodeValueCodec;
+import com.linecorp.armeria.common.zookeeper.NodeValueCodec;
+import com.linecorp.armeria.common.zookeeper.ZKConnector;
+import com.linecorp.armeria.common.zookeeper.ZooKeeperException;
+
+/**
+ * A ZooKeeper-based {@link EndpointGroup} implementation. This {@link EndpointGroup} retrieves the list of
+ * {@link Endpoint}s from a ZooKeeper zNode's all children node's value and updates it when the children of the
+ * zNode changes. When a ZooKeeper session expires, it will automatically reconnect to the ZooKeeper with
+ * exponential retry delay,starting from 1 second up to 60 seconds.
+ */
+public class NodeChildEndpointGroup extends ZKConnector implements EndpointGroup {
```

Comment:
@trustin I'm afraid you have the point. Leaving the `ZKConnector` as a public class give the API users an unnecessary `class` view, which will lead to misuse. In fact it is just a class        
 for inside-module using. But since it has some features need to be defined by subclasses, 
 class composition could not resolve this, any better ideas?